flipt-io/flipt-node
# feat: flipt-node cli

Proof of concept for CLI tool to convert export to types. Follow-up on #6 and https://github.com/flipt-io/flipt/pull/1849)

This uses the `typescript` factory functions to build an AST and generate valid TS this way.

How to test:

- Export from flipt or use example from [here](https://www.flipt.io/docs/experimental/filesystem-backends#defining-flag-state) to make a features.yml file
- `yarn build`
- `cat features.yml | node ./dist/cli`
(After publishing on NPM, this can be used like `cat features.yml | npx @flipt-io/flipt`)

Output should look like this:

```ts
// Generated by flipt-node Thu Jul 13 2023 16:09:10 GMT+0900 (Korean Standard Time)
export type Flags = {
    key: "awesomeNewFeature";
    value: "enabled" | "disabled";
} | {
    key: "awesomeNewFeature2";
    value: unknown;
};
export type Context = {
    organization: string;
    age: number;
};
```

This file can be put in a TS project and used in any way the user deems useful. For example, you can extract all FlagKeys with `type FlagKey = Flags['key']` and make stricter API requests/responses like this:

```ts
import { EvaluationRequest, EvaluationResponse } from '@flipt-io/flipt/api';
import { Flags, Context } from './generated/flipt.ts';

type FlagKey = Flags['key'];

export interface StrictEvaluationRequest<F extends FlagKey = FlagKey> extends EvaluationRequest {
  flagKey: F;
  context: Context;
}

export interface StrictEvaluationResponse<F extends FlagKey = FlagKey> extends EvaluationResponse {
  value: (Flags & { key: F })['value']
}
```
